File: leetcode/go/1043.partition-array-for-maximum-sum.go
func maxSumAfterPartitioning(arr []int, k int) int {
    k++
    n := len(arr)
    dp := make([]int, k)
    end, curr_max := 0, 0
    for start:=n-1; start > -1 ; start-- {
        curr_max = 0
        end = min(start + k - 1, n)
        for i:=start; i<end ; i++ {
            curr_max = max(curr_max, arr[i])
            dp[start % k] = max(
                dp[start % k], 
                dp[(i+1) % k] + (curr_max * (i-start+1)))
        }
    }
    return dp[0]
}

func min(a int, b int) int {
    if a < b {return a}
    return b
}

func max(a int, b int) int {
    if a > b {return a}
    return b
}

File: leetcode/rust-leetcode/src/1043.partition-array-for-maximum-sum.rs
impl Solution {
    pub fn max_sum_after_partitioning(arr: Vec<i32>, k: i32) -> i32 {
        let k = (k+1) as usize;
        let mut dp = vec![0; k];
        let n = arr.len();
        let mut cmax = 0;
        let mut end = 0;
        for start in (0..n).rev() {
            cmax = 0;
            end = n.min(k-1+start);
            for i in start..end {
                cmax = cmax.max(arr[i]);
                dp[start % k] = dp[start % k].max(dp[(i+1) % k] + (cmax * ((i-start+1) as i32)));
            }
        }
        dp[0]
    }
}
